documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+# Summary
+A new API for generic maths in Silk.NET.Maths 3.X, leveraging modern .NET features such as `INumber<T>` and vectorization.
+
+This API aims to replace the existing implementation of Silk.NET.Maths.
+
+# Contributors
+- Maxine H (@uwx)
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+- Should replace the existing APIs without too much growing pains
+- Should try to roughly match System.Numerics APIs, with concessions for design oversights in those APIs
+- Should use vectorized implementations for methods when possible
+- Prefer generic specialization and extension methods, and keep the type of `T` as wide as possible in the struct itself
+- Primitives are read-only, `with` preferred for modification
+- Public fields instead of properties
+- Implement the following interfaces:
+    - IEquatable
+    - IEquatable to equivalent System.Numerics primitive, if available
+    - IReadOnlyList
+    - IUtf8SpanFormattable
+    - IUtf8SpanParsable
+- Provide explicit implementations of the following interfaces:
+    - INumberBase (except for methods which are relevant to vectors)
```

Comment:
I would recommend against implementing `INumberBase`. That type is primarily intended for "scalar" like values and has APIs on it that effectively assume such a shape.

Vectors are not, in the programmatic sense, numbers. Rather they are most closely viewed as a collection of numbers.

The "better" approach would be to define some new `interface IVector<TSelf, T>` and then do `where TSelf : IVector<TSelf, T>` and `where T : INumberBase<T>` (although I personally still hold the opinion that vectors fundamentally need to be floating-point and that will likely be how the BCL follows suite for their generic `Vector2<T>` and related APIs).

Reply:
Noted. If we are to not implement INumberBase however, we need to have a discussion about what members are actually relevant to copy over.